### ISSUE TITLE
[Docs] Adopt Versioning

### ISSUE THREAD

**Issue by robscott** [kubernetes-sigs/gateway-api#3115]
**What would you like to be added**:
Adopting some form of versioning for documentation would enable us to review and merge docs for a feature before it is released. This may just be versioned branches, or selectively versioning sections of the website. Unfortunately different areas of our website could use different approaches to versioning:

| Area | Goal |
| - | - |
| Implementations | Always Latest |
| Conformance Results/Tables | Likely want to wait one month after a release before displaying latest, also want to be able to show historical results |
| GEPs | Likely want to who latest status for ease of reference, but this could get confusing as GEP status on website may not match released status |
| Features + Guides | Should match release |
| API Spec | Should match release |
| Contributing | Always Latest |

Some potential options:

**1. Reuse release branches**
With this approach, we'd pin the docs website to the latest release-1.x branch. Every fix, update, and GEP would need to be cherry picked to that branch. Future-looking updates would go straight to main.

**2. Introduce new docs branches**
As part of a release process, we'd create a docs branch reserved for future releases, and send unreleased updates directly to that. I don't think this would allow us to avoid unreleased updates to the API spec documentation though.

**3. Introduce selective versioning for some subset of the website**
Instead of using branches, we could duplicate sections of our website that would benefit from versioning, and continue to serve everything from main. 

**Why this is needed**:
In https://github.com/kubernetes-sigs/gateway-api/pull/3108, we have a feature that has met all graduation criteria for standard channel in v1.2, but we don't really have a good place to put the corresponding docs updates before v1.2 is formally released. This is just one of many examples of a common pain point when documenting features in Gateway API.

**Comment by youngnick** [comment]
I think that 1 is pretty unavoidable - although moving to Hugo instead of mkdocs would allow us to use the same tooling as upstream Kubernetes, which may make getting contributors easier. That's a big job though.

**Comment by xtineskim** [comment]
for displaying conformance results, is there consensus on wanting it to be 1 month post release, or the suggested wait till 3 implementations are submitted ([comment](https://github.com/kubernetes-sigs/gateway-api/pull/3088#discussion_r1609383547))?

**Comment by robscott** [comment]
> for displaying conformance results, is there consensus on wanting it to be 1 month post release, or the suggested wait till 3 implementations are submitted ([comment](https://github.com/kubernetes-sigs/gateway-api/pull/3088#discussion_r1609383547))?

Maybe instead of "whichever comes first", we should say "whichever comes later", so we can ensure that implementations always have at least 1 month to implement the latest release of an API to be included in our conformance table.

**Comment by youngnick** [comment]
Agreed, I like "one month after release, or whenever three implementations support it, whichever is longer" as the rule.

**Comment by guicassolato** [comment]
_Prior art:_ [docs.kuadrant.io](https://docs.kuadrant.io/) uses mkdocs as well and it has versioned docs.

I think it's based on https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning IIRC.
